src/toil/server/wes/toil_backend.py

```diff
@@ -0,0 +1,318 @@
+
+    def set_up_run(self) -> None:
+        """ Set up necessary directories for the run."""
+        if not os.path.exists(self.exec_dir):
+            os.makedirs(self.exec_dir)
+
+    def clean_up(self) -> None:
+        """ Clean directory and files related to the run."""
+        shutil.rmtree(os.path.join(self.work_dir))
+
+    def queue_run(self, request: Dict[str, Any], options: List[str]) -> None:
+        """This workflow should be ready to run. Hand this to Celery."""
+        with open(os.path.join(self.work_dir, "request.json"), "w") as f:
+            json.dump(request, f)
+
+        run_wes.apply_async(args=(self.work_dir, request, options),
+                            task_id=self.run_id,  # set the Celery task ID the same as our run ID
+                            ignore_result=True)
+
+    def get_output_files(self) -> Any:
+        """
+        Return a collection of output files that this workflow generated.
+        """
+        return json.loads(self.fetch("outputs.json", "{}"))
+
+
+class ToilBackend(WESBackend):
+    """
+    WES backend implemented for Toil to run CWL, WDL, or Toil workflows. This
+    class is responsible for validating and executing submitted workflows.
+
+    Single machine implementation -
+    Use Celery as the task queue and interact with the "workflows/" directory
+    in the filesystem to store and retrieve data associated with the runs.
+    """
+
+    def __init__(self, work_dir: str, options: List[str]) -> None:
+        super(ToilBackend, self).__init__(options)
+        self.work_dir = work_dir
+        self.supported_versions: Dict[str, List[str]] = {}
+
+    def register_wf_type(self, name: str, supported_versions: List[str]) -> None:
+        """
+        Register a workflow type that this backend should support.
+        """
+        self.supported_versions[name.lower()] = supported_versions
+
+    def _get_run(self, run_id: str, should_exists: Optional[bool] = None) -> ToilWorkflow:
+        """
+        Helper method to instantiate a ToilWorkflow object.
+
+        :param run_id: The run ID.
+        :param should_exists: If set, ensures that the workflow run exists (or
+                              does not exist) according to the value.
+        """
+        run = ToilWorkflow(run_id, work_dir=os.path.join(self.work_dir, run_id))
+
+        if should_exists is not None:
+            if should_exists and not run.exists():
+                raise WorkflowNotFoundException
+            if should_exists is False and run.exists():
+                raise WorkflowConflictException(run_id)
+
+        return run
+
+    def get_runs(self) -> Generator[Tuple[str, str], None, None]:
+        """ A generator of a list of run ids and their state."""
+        if not os.path.exists(self.work_dir):
+            return
+
+        for run_id in os.listdir(self.work_dir):
+            run = self._get_run(run_id)
+            if run.exists():
+                yield run_id, run.get_state()
```

Comment:
Is it possible for a run to leave Celery without leaving the `work_dir` (for example, if the server reboots)? Is it ever going to be noticed and cleaned up, or does the user need to clean out the existing work_dir contents before starting the server at system boot?

Reply:
If celery or the WES server didn't stop gracefully (e.g.: crashes, sigkills, server reboots), it is possible that certain runs get left behind and would require manual clean up. If we want, we could check and set the runs that are not in a finished state to "EXECUTOR_ERROR" or something before we start the server.

Reply:
The "executor" in the context of TES and WES is I think the actual submitted work to be run. So `SYSTEM_ERROR` might be the right choice for workflows that Celery loses.

It might make sense to `--restart` these workflows if we discover them, though.